jena-core/src/main/java/org/apache/jena/util/PrefixMappingUtils.java

```diff
@@ -0,0 +1,362 @@
+        Trie<String> trie = new Trie<>() ; 
+        // Change this to "add(uri, uri)" to calculate the uris. 
+        pmap.forEach((prefix,uri)-> trie.add(uri, prefix)) ;
+        Iterator<Triple> iter = graph.find(null, null, null) ;
+        // Prefixes in use.
+        // (URIs if "add(uri, uri)")
+        Set<String> inUse = new HashSet<>() ;
+        
+        // Process to apply to each node
+        // Accumulate any prefixes into 'inUse' if the data URI
+        // is partially matched by a prefix URIs in the trie.  
+        Consumer<Node> process = (node)->{
+            if ( ! node.isURI() )
+                return ;
+            String uri = node.getURI() ;
+            // Get all prefixes whose URIs are candidates 
+            List<String> hits = trie.partialSearch(uri) ;
+            if ( hits.isEmpty() )
+                return ;
+            inUse.addAll(hits) ;
+        } ;
+        
+        while(iter.hasNext()) {
+            Triple triple = iter.next() ;
+            process.accept(triple.getSubject()) ;
+            process.accept(triple.getPredicate()) ;
+            process.accept(triple.getObject()) ; 
+            if ( pmap.size() == inUse.size() )
+                break ;
+        }
+        
+        if ( pmap.size() == inUse.size() )
+            return prefixMapping ;
+
+        // Build result.
+        PrefixMapping pmap2 = new PrefixMappingImpl() ;
+        inUse.forEach((prefix)-> pmap2.setNsPrefix(prefix, prefixMapping.getNsPrefixURI(prefix)) ) ;
+        return pmap2 ;
+    }
+
+    /**
+     * Analyse the graph to see which prefixes of the graph are in use.
+     * <p>
+     * This function attempts to process each URI in the graph as if it were to be printed
+     * in Turtle. Only prefixes that lead to valid output strings are returned. This is
+     * more expensive than {@link #calcInUsePrefixMapping(Graph)}.
+     * <p>
+     * This function does not calculate new prefixes.
+     * 
+     * @see #calcInUsePrefixMappingTTL(Graph)
+     */
+    public static PrefixMapping calcInUsePrefixMappingTTL(Graph graph) {
+        PrefixMapping prefixMapping = graph.getPrefixMapping() ;
+        if ( prefixMapping == null )
+            return null ;
+        return calcInUsePrefixMappingTTL(graph, prefixMapping) ;
+    }
+    
+    /**
+     * Analyse the graph to see which prefixes of the given {@link PrefixMapping} are used
+     * by the graph triples.
+     * <p>
+     * This function attempts to process each URI in the graph as if it were to be printed
+     * in Turtle. Only prefixes that lead to valid output strings are returned. This is
+     * more expensive than {@link #calcInUsePrefixMapping(Graph, PrefixMapping)}.
+     * <p>
+     * This function does not calculate new prefixes.
+     * 
+     * @see #calcInUsePrefixMapping(Graph, PrefixMapping)
+     */
+    public static PrefixMapping calcInUsePrefixMappingTTL(Graph graph, PrefixMapping prefixMapping) {        
```

Comment:
Maybe leave a TODO to factor out common logic between this and `calcInUsePrefixMapping`?